Framework/PythonInterface/mantid/dataobjects/src/Exports/Workspace2D.cpp

```diff
@@ -1,19 +1,174 @@
 #include "MantidDataObjects/Workspace2D.h"
+
+#include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/Axis.h"
+#include "MantidAPI/SpectrumInfo.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidIndexing/IndexInfo.h"
+#include "MantidIndexing/SpectrumNumber.h"
+#include "MantidKernel/Logger.h"
+#include "MantidKernel/OptionalBool.h"
+#include "MantidKernel/Unit.h"
+#include "MantidPythonInterface/kernel/Converters/CloneToNumpy.h"
+#include "MantidPythonInterface/kernel/Converters/WrapWithNumpy.h"
+#include "MantidPythonInterface/kernel/Converters/VectorToNDArray.h"
+#include "MantidPythonInterface/kernel/Converters/NDArrayToVector.h"
 #include "MantidPythonInterface/kernel/GetPointer.h"
 #include "MantidPythonInterface/kernel/Registry/RegisterWorkspacePtrToPython.h"
+#include "MantidTypes/SpectrumDefinition.h"
 
 #include <boost/python/class.hpp>
+#include <boost/python/dict.hpp>
+#include <boost/python/list.hpp>
+#include <boost/python/make_constructor.hpp>
+#include <boost/python/tuple.hpp>
 
 using Mantid::API::MatrixWorkspace;
 using Mantid::DataObjects::Workspace2D;
+using Mantid::SpectrumDefinition;
 using namespace Mantid::PythonInterface::Registry;
+using namespace Mantid::Indexing;
 using namespace boost::python;
 
 GET_POINTER_SPECIALIZATION(Workspace2D)
 
+class Workspace2DPickleSuite : public boost::python::pickle_suite {
+public:
+  static dict getstate(const Workspace2D &ws) {
+    using namespace Mantid::PythonInterface::Converters;
+
+    // Python lists that will be added to dict
+    list spectraList;
+    list errorList;
+    list binEdgeList;
+    list detectorList;
+    list specNumList;
+
+    auto spectra = ws.getNumberHistograms();
+    const auto &indexInfo = ws.indexInfo();
+    const auto &spectrumDefinitions = indexInfo.spectrumDefinitions();
+
+    for (decltype(spectra) i = 0; i < spectra; ++i) {
+      const auto &histo = ws.histogram(i);
+
+      const auto &spectraData = histo.counts().data().rawData();
+      const auto &errorData = histo.countStandardDeviations().rawData();
+      const auto &binEdges = histo.binEdges().rawData();
+
+      spectraList.append(object(
+          handle<>(VectorToNDArray<double, WrapReadOnly>()(spectraData))));
+      errorList.append(
+          object(handle<>(VectorToNDArray<double, WrapReadOnly>()(errorData))));
+      binEdgeList.append(
+          object(handle<>(VectorToNDArray<double, WrapReadOnly>()(binEdges))));
+
+      auto spectrumNumber = indexInfo.spectrumNumber(i);
+      const auto &spectrumDefinition = (*spectrumDefinitions)[i];
+      std::vector<size_t> detectorIndices;
+
+      for (size_t j = 0; j < spectrumDefinition.size(); ++j) {
+        size_t detectorIndex = spectrumDefinition[j].first;
+        detectorIndices.emplace_back(std::move(detectorIndex));
```

Comment:
Not supporting workspaces with scanning instrument here, does it make sense to check somewhere and fail? Otherwise you can end up in an inconsistent state.

Reply:
Agreed. That should be checked.